feat(codegen): angular element support 244

devinmotion opened the pull request:
Add Angular Element code generation support.
The solution tries to streamline the process described in `ngx-build-plus` by Manfred Steyer (https://github.com/manfredsteyer/ngx-build-plus).
If you generate the assets with the editor and follow the README you will create a minimal toolchain to create Angular Elements in general.

Closes #244

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature (**codegen**)
- [x] Code style update (formatting, local variables) -> Whitespace in `codegen.service.ts`
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying or link to a relevant issue. -->

Issue Number: #244


## What is the new behavior?


## Does this PR introduce a breaking change?



- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

-  [x] add Component codegen
 - [x]  add spec file codegen (if applicable)
 - [x]  add HTML codegen
 - [x]  add styles codegen
 - [x]  add readme codegen
 - [ ]  add export to stackblitz (or other external editor) -> May not make sense because the generation needs manual steps afterwards at the moment.
 - [x]  add unit test -> However it's very basic!
 - [x]  add entry to the editor menu

devinmotion commented:
@manekinekko @Jefiozie Could you please have a look whether this Angular Element codegen approach would fit your needs? It basically does some streamlining to the process described in `ngx-build-plus` (https://github.com/manfredsteyer/ngx-build-plus).

Assuming it is ok, you may help me by getting everything ready for the final PR:
- Are there any additional docs that need to be updated for the Angular Element support?
- I have minimal unit testing support. Are there any e2e tests or unit tests that would make sense to add here?
- I will squash and reword the commit in order to align with the defined commit message format. Is that ok and the typical the way to go?

manekinekko commented:
This is awesome!!!! I'm gonna review it asap 👍

manekinekko commented:
@devinmotion @davidwuerfel The PR looks good. You have done an incredible job! Thank you so much! Everything is working as expected. 

@devinmotion no worries about the tests and e2e. We are still thinking about the best strategy for testing the custom code generators. Any suggestions?

manekinekko commented:
@Shikanime can we get #303 reviewed and merged asap so we can merge this PR after and have it ready for next release?

Jefiozie commented:
@devinmotion @davidwuerfel,  great work guys on this PR, it looks good!! Thank you for working on this PR 🙏🏻

devinmotion commented:
> @devinmotion @davidwuerfel, great work guys on this PR, it looks good!! Thank you for working on this PR 🙏🏻

Thank you ❤️ , you're really welcome! Had a lot of fun, it was a pleasure. Ready to do some more in the future 🚀. Actually *devinmtion* is my private account whereas *davidwuerfel* is my company account. A bit confusing, depending on the mail adress that has been configured in git on different machines. I will soon change/transfer from *devinmotion* (which is really an old relict) to something that aligns with my Twitter handle. Prepare to be even more confused! 😆

devinmotion commented:
I'll have a look into merging the current master back into this branch soon.

devinmotion commented:
@manekinekko @Jefiozie I merged current master, rebased, squashed and aligned to your commit message format. Should be ready to review (again 😁) and to be merged into master.

Some remarks while merging the current master:
- `make` does not work on Windows out-of-the-box. I tried to use an [npm-library](https://www.npmjs.com/package/make), however multiple npm run lines in the make file did not work. Had no solution for this atm, and ran the scripts by hand.
- Are you using any formatters? I missed a project-wide `.prettierrc`.
- Some refactorings in the new codegen approach result in any-typed parameters. Are you planning to add types back again?

If you would like to fix some of these, could I help creating some issues?

manekinekko commented:
awesome! Thank you for rebasing and thank you for the feedback. 
The other codegen redactor was a huge work and we might have missed some small changes. Will plan this rework for our next sync, next week.
